### ISSUE TITLE
With small Y value, the rod becomes square

### ISSUE THREAD

**Issue by yintaibing** [imaNNeoFighT/fl_chart#147]
`fl_chart` version: `0.6.0`

When the Y value is small, the rod is always a square. I found some codes below in `bar_chart_painter.dart`.
```
void drawBars(Canvas canvas, Size viewSize, List<GroupBarsPosition> groupBarsPosition) {
  //...
  final yFrom = getPixelY(0, drawSize);
  from = Offset(
    x,
    yFrom - roundedRadius,
  );

  var yTo = getPixelY(barRod.y, drawSize);
  // this is the reason
  if ((yFrom - yTo).abs() <= barRod.width) {
    yTo = yFrom - barRod.width;
  }
  //...
}
```

Why don't you allow the rod's height is smaller than its width? Can you remove this restriction? Or provide some input args to customize it? Thank you very much.

submitted a [PR#148](https://github.com/imaNNeoFighT/fl_chart/pull/148)

**Comment by imaNNeoFighT** [comment]
That is a mistake, 
we have an `isRount` property, and we should do this only when `isRound` is true.
I will let you know what to do in the PR.
Thanks for it :)

**Comment by imaNNeoFighT** [comment]
And for sure sorry for the delay.

**Comment by juumixx** [comment]
I have another suggestion how to solve this issue:
Currently, the rods are round or square, but I would like to have a configurable edge radius.
It would require an additional property like `cornerRadius`, which would be applied only to the upper edge of the bar. When not set (`cornerRadius=null`), it would behave like it is now for the best compatibility, but instead of setting `isRound=false` when you want totally square bars, you would have to set `cornerRadius=0`, which would also be the minimum height for the bar.
PR coming soon.

**Comment by imaNNeoFighT** [comment]
Great solution, I'm totally with you.
but I think we should let the user set radius for `topLeft`, `topRight`, `bottomLeft`, `bottomRight`  of the bar (not only top).
I think we can use the [BorderRadius](https://api.flutter.dev/flutter/painting/BorderRadius-class.html) class for the sake of consistency.

**Comment by juumixx** [comment]
You are absolutely right. I will use the BorderRadius it in my change.

**Comment by imaNNeoFighT** [comment]
Waiting for your PR.
Thanks.

**Comment by juumixx** [comment]
I pushed my implementation to https://github.com/juumixx/fl_chart/commit/aa0c62f880f1565362adbe6e4f64b947b3f096fe

I also used the rounded rectangles in the stack items, but it creates some weird behavior you can see in this screenshot:
![](https://github.com/juumixx/fl_chart/raw/aa0c62f880f1565362adbe6e4f64b947b3f096fe/repo_files/images/bar_chart/rounded_bars.jpg)

* The first bar is rounded top & bottom, but it looks somehow weird when the stack items are drawn like this.
* The second bar draws all stack items from the bottom (see https://github.com/juumixx/fl_chart/commit/aa0c62f880f1565362adbe6e4f64b947b3f096fe#diff-0adf96074192401bef1b202d80de17c7R137 ff). This looks much better, but the stack items have to be drawn from the bottom up and from back to front. I don't like the complexity it adds to the way the library has to be used (order of stack items matter, overlap of stack items matter).

I could also rewrite the way the stack items are drawn and only color that part of the bar (so the stack items would not be affected by the rounding of the bar). I think this would be the most understandable way to do it. What do you think?

Btw. this issue is fixed by this code. It always uses the radius of the bar as a minimum height for the bar. When the radius is 0, the bar won't be square anymore but always represents the actual height of the bar's value. When the radius greater than 0, the bar will always be at least as high as the radius (except when the value is 0, then the bar won't be drawn). With a border radius `null` the radius defaults to the half width of the bar (which equals to the `isRound=true` of the previous implementation.

**Comment by juumixx** [comment]
So a problem I stumbled upon is how this part between the bottom of the graph and the start of the full-width-bar should behave:
![0problem](https://user-images.githubusercontent.com/5200179/71907493-71946100-316c-11ea-94c0-c822e85f7fc0.png)

* Either every part of the bar always extend over their actual value by the radius of the rounded corners. This would mean we would have to recalculate the heights depending on the radius, because 0 is now at the `?` in the image and not at the bottom of the chart anymore.
* Otherwise, we cannot guarantee that the rounding is not cut off, which would result in bad looks in some cases. Try out 0.1 as value.

**Comment by juumixx** [comment]
My preferred solution would be to use clipping to draw the stack elements, which would also allow to draw stack parts which are smaller than the rounding, while not having to care about visual issues and still preserving the correctness of the y axis:
![clip](https://user-images.githubusercontent.com/5200179/71909320-16646d80-3170-11ea-8c34-9c3132332954.png)
As you can see in this example, there is a stack element on top of the left bar, which only spans about half of the rounded corners.

I pushed the implementation of it here:
https://github.com/juumixx/fl_chart/commit/a27a4e07de8069cd1a376423e7e98f232c8535be

**Comment by imaNNeoFighT** [comment]
@juumixx 
why you are trying to add the radius to each stack section?
I think we can apply it just in each rod. and it satisfies all of us.
In the stacked bar chart, we draw all of the stack sections rectangular (then the height of each section is exact same value as provided), then we cut the area out of the shape with the provided radius.
check this code:
https://github.com/imaNNeoFighT/fl_chart/blob/18d9a69d2b07c6a219a83dc8455c56f6e1202706/lib/src/chart/bar_chart/bar_chart_painter.dart#L262 
Keep it simple man, just apply the radius on only each Rod.
Thanks!

**Comment by juumixx** [comment]
I just wanted to mimic the previous behavior with `isRound=true`, where each part of the bar was drawn as a single line with rounded edges. But you are totally right, let's keep it simple.
I'll update the documentation and create a PR.

**Comment by imaNNeoFighT** [comment]
Fixed in [0.7.0](https://github.com/imaNNeoFighT/fl_chart/releases/tag/0.7.0)
